pkg/api/components.go

```diff
@@ -0,0 +1,257 @@
+		Containers: []KubernetesContainerSpec{
+			{
+				Name:  common.APIServerComponentName,
+				Image: getComponentDefaultContainerImage(common.APIServerComponentName, cs),
+			},
+		},
+		Config: map[string]string{
+			"command": getAPIServerDefaultCommandString(cs),
+		},
+	}
+
+	defaultAddonManagerComponentConfig := KubernetesComponent{
+		Name:    common.AddonManagerComponentName,
+		Enabled: to.BoolPtr(true),
+		Containers: []KubernetesContainerSpec{
+			{
+				Name:  common.AddonManagerComponentName,
+				Image: getComponentDefaultContainerImage(common.AddonManagerComponentName, cs),
+			},
+		},
+	}
+
+	defaultComponents := []KubernetesComponent{
+		defaultSchedulerComponentConfig,
+		defaultControllerManagerComponentConfig,
+		defaultCloudControllerManagerComponentConfig,
+		defaultAPIServerComponentConfig,
+		defaultAddonManagerComponentConfig,
+	}
+	// Add default component specification, if no user-provided spec exists
+	if kubernetesConfig.Components == nil {
+		kubernetesConfig.Components = defaultComponents
+	} else {
+		for _, component := range defaultComponents {
+			kubernetesConfig.Components = appendComponentIfNotPresent(kubernetesConfig.Components, component)
+		}
+	}
+
+	for _, component := range defaultComponents {
+		synthesizeComponentsConfig(kubernetesConfig.Components, component, isUpgrade)
+	}
+}
+
+func appendComponentIfNotPresent(components []KubernetesComponent, component KubernetesComponent) []KubernetesComponent {
+	if component.Name != "" {
+		i := getComponentsIndexByName(components, component.Name)
+		if i < 0 {
+			return append(components, component)
+		}
+	}
+	return components
+}
+
+func getComponentsIndexByName(components []KubernetesComponent, name string) int {
+	for i := range components {
+		if components[i].Name == name {
+			return i
+		}
+	}
+	return -1
+}
+
+// assignDefaultComponentVals will assign default values to component from defaults, for each property in component that has a zero value
+func assignDefaultComponentVals(component, defaultComponent KubernetesComponent, isUpgrade bool) KubernetesComponent {
+	if component.Name == "" {
+		component.Name = defaultComponent.Name
+	}
+	if component.Enabled == nil {
+		component.Enabled = defaultComponent.Enabled
+	}
+	if !to.Bool(component.Enabled) {
+		return KubernetesComponent{
+			Name:    component.Name,
+			Enabled: component.Enabled,
+		}
+	}
+	if component.Data != "" {
+		return KubernetesComponent{
+			Name:    component.Name,
+			Enabled: component.Enabled,
+			Data:    component.Data,
+		}
+	}
+	for i := range defaultComponent.Containers {
+		c := component.GetContainersIndexByName(defaultComponent.Containers[i].Name)
+		if c < 0 {
+			component.Containers = append(component.Containers, defaultComponent.Containers[i])
+		} else {
+			if component.Containers[c].Image == "" || isUpgrade {
+				component.Containers[c].Image = defaultComponent.Containers[i].Image
+			}
+			if component.Containers[c].CPURequests == "" {
```

Comment:
FYI this configurable resource limits functionality is not yet implemented, but it seems likely that it may be at some point in the future, so it was optimal to functionally implement this as part of the data spec now (w/ tests), during this initial abstraction introduction. If we do want it later, it saves time having already done this foundational work; in the meantime it has no effect on any of the currently implemented component templates (and it won't be documented).